### ISSUE TITLE
[Bug] dbt deps does not install the packages of dependent local projects

### ISSUE THREAD

**Issue by joshuajohananorthos** [dbt-labs/dbt-core#9738]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

We are using dbt-core 1.6 and I am testing upgrading to 1.7.

The issue I am seeing now is:
`'dbt_utils' is undefined. This can happen when calling a macro that does not exist. Check for typos and/or install package dependencies with "dbt deps".`

I can see the source of the problem. When I run dbt deps for the project, it only installs the packages for the current project. The top level project references another local project in which that references dbt_utils.

So it looks like this:
client_project -- local package in packages.yml --> abstract_data_model_project -- dbt hub reference in packages.yml --> dbt_utils

I also tried to changing the files to dependencies.yml to see if that changed anything and it did not.

### Expected Behavior

We are using dbt-snowflake==1.6.0 which installs dbt-core 1.6.10. This version of works and installs all the packages of the child project into the root projects `dbt_packages` directory.

### Steps To Reproduce

1. Create a project and add the `dbt_utils` package to it.
2. Create another project and add the first project as a local package.
3. Run `dbt deps` from the second package
4. `dbt_utils` will not be installed, but local package will be installed.

### Relevant log output

```shell
Command and output from the root package. With 1.6.x this would also install dbt_utils from dataview_package in this example.

$ dbt deps
19:35:42  Running with dbt=1.7.9
19:35:42  Installing ../../modules/dataview_package
19:35:42  Installed from <local @ ../../modules/dataview_package>
19:35:42  Installing ../../modules/utils
19:35:42  Installed from <local @ ../../modules/utils>
```

### Environment

```markdown
- OS:Ubuntu 22.04
- Python:3.9.16
- dbt:
Core:
  - installed: 1.7.9
  - latest:    1.7.9 - Up to date!

Plugins:
  - snowflake: 1.7.2 - Up to date!
```

### Which database adapter are you using with dbt?

snowflake

**Comment by dbeatty10** [comment]
Thanks for raising this issue @joshuajohananorthos.

Does your project look like the following example?

If so, I wasn't able to reproduce the issue you reported. Could you try out this example and see if it works for you or not? If not, can you try to tweak it to exhibit the issue you experienced?

Here's the way my folders and files look:
```
.
├── modules
│   └── dataview_package
│       ├── dbt_project.yml
│       └── packages.yml
└── projects
    └── my_project
        ├── dbt_project.yml
        └── packages.yml
```

`modules/dataview_package/packages.yml`

```
packages:
  - package: dbt-labs/dbt_utils
    version: 1.1.1
```

`projects/my_project/packages.yml`

```yaml
packages:
  - local: ../../modules/dataview_package
```

### This worked

```shell
dbt deps --project-dir projects/my_project 
```

Output:
```
$ dbt deps --project-dir projects/my_project 
00:16:52  Running with dbt=1.7.8
00:16:56  Installing ../../modules/dataview_package
00:16:56  Installed from <local @ ../../modules/dataview_package>
00:16:56  Installing dbt-labs/dbt_utils
00:16:59  Installed from version 1.1.1
00:16:59  Up to date!
```

### And so did this

```shell
cd projects/my_project
dbt deps
```

Output:
```
$ dbt deps
00:07:55  Running with dbt=1.7.8
00:07:58  Installing ../../modules/dataview_package
00:07:58  Installed from <local @ ../../modules/dataview_package>
00:07:58  Installing dbt-labs/dbt_utils
00:08:00  Installed from version 1.1.1
00:08:00  Up to date!
```

**Comment by joshuajohananorthos** [comment]
@dbeatty10 Thanks for putting together an example to test. I started fresh and slowly built up your example and everything worked for me. Then I tried to add more to better line up with the actual projects and I still could not get `dbt deps` to fail.

At this point I went back to the actual projects and created an entirely new virtual env and built it up starting with dbt. This worked fine and I was not able to reproduce.

Ultimately it was some mix of the packages I had installed and then upgrading dbt that lead to this issue. I was not able to reproduce from a clean environment.

What is really weird is that the issue first appeared in CI pipeline. I created the PR and then ran our tests to make sure everything would work and this issue popped up. Then I tried it locally and was able to duplicate the problem. I do not still have the CI worker as we spin them down, but we keep them up most of the day and we don't *completely* clean up the environment between runs. I created a perfect storm of issues that lead to this problem.

I do have the output of `pip freeze` from my local environment if you want it. 

Thanks for your help and the steps I have taken were steps I should have taken before I opened the issue.
